### PR TITLE
feat(evm): add push{i} opcodes

### PR DESCRIPTION
**Pull request by LucasLvy**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #194 

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/evm/src/instructions/push_operations.cairo [hunk 1]
```diff
@@ -1,227 +1,272 @@
...
+    }
+}
+
+#[generate_trait]
+impl PushOperations of PushOperationsTrait {
+    /// 5F - PUSH0 operation
+    /// # Specification: https://www.evm.codes/#5f?fork=shanghai
+    #[inline(always)]
+    fn exec_push0(ref self: ExecutionContext) -> Result<(), EVMError> {
+        internal::exec_push_i(ref self, 0)
```
Comment (Eikix) [inline crates/evm/src/instructions/push_operations.cairo:28]: is this a special case? as in: it pushes 0 to the stack without reading code

File: crates/evm/src/instructions/push_operations.cairo [hunk 2]
```diff
@@ -1,227 +1,272 @@
...
-
-/// 0x7F - PUSH32 operation
-/// # Specification: https://www.evm.codes/#7f?fork=shanghai
-fn exec_push32(ref context: ExecutionContext) {
-    internal::exec_push_i(ref context, 32);
+    #[inline(always)]
+    fn exec_push_i(ref context: ExecutionContext, i: u8) -> Result<(), EVMError> {
+        let data = context.read_code(i.into());
+        let mut i: usize = i.into();
+        context.stack.push(load_word(i, data))
```
Comment (Eikix) [inline crates/evm/src/instructions/push_operations.cairo:18]: does load word just load i bytes from data?

File: crates/evm/src/instructions/push_operations.cairo [hunk 3]
```diff
@@ -1,227 +1,272 @@
...
-}
-
-
-/// 0x7F - PUSH32 operation
-/// # Specification: https://www.evm.codes/#7f?fork=shanghai
-fn exec_push32(ref context: ExecutionContext) {
-    internal::exec_push_i(ref context, 32);
+    #[inline(always)]
+    fn exec_push_i(ref context: ExecutionContext, i: u8) -> Result<(), EVMError> {
+        let data = context.read_code(i.into());
```
Comment (Eikix) [inline crates/evm/src/instructions/push_operations.cairo:16]: does read code return a i bytes slice from code?

**Review by Eikix — COMMENTED** [review]
some questions

File: crates/evm/src/instructions/push_operations.cairo (hunk 1 above)
Comment (LucasLvy) [reply crates/evm/src/instructions/push_operations.cairo:28]: this is exactly what it does yes. From evm.codes `Place value 0 on stack`

File: crates/evm/src/instructions/push_operations.cairo (hunk 2 above)
Comment (LucasLvy) [reply crates/evm/src/instructions/push_operations.cairo:18]:
```rs
/// Loads a sequence of bytes into a single u128 in big-endian
///
/// # Arguments
/// * `len` - The number of bytes to load
/// * `words` - The bytes to load
///
/// # Returns
/// The packed u256
fn load_word(mut len: usize, words: Span<u8>) -> u256 {
  ```

File: crates/evm/src/instructions/push_operations.cairo (hunk 3 above)
Comment (LucasLvy) [reply crates/evm/src/instructions/push_operations.cairo:16]:
```rs
/// Reads and return data from bytecode.
/// The program counter is incremented accordingly.
///
/// # Arguments
///
/// * `self` - The `ExecutionContext` instance to read the data from.
/// * `len` - The length of the data to read from the bytecode.
#[inline(always)]
fn read_code(ref self: ExecutionContext, len: usize) -> Span<u8> {
```

File: crates/evm/src/interpreter.cairo [hunk 4]
```diff
@@ -343,135 +343,135 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         }
         if opcode == 95 {
             // PUSH0
-            push_operations::exec_push0(ref context);
+            context.exec_push0()?;
```
Comment (enitrat) [inline crates/evm/src/interpreter.cairo:346]: all of these should be `context.exec_opcode()?`
 (maybe with an explicit return) to propagate eventual errors

Comment (enitrat) [reply crates/evm/src/interpreter.cairo:346]: cc @lambda-0x can you add this to your PR? then @LucasLvy you will be able to rebase once I merge @lambda-0x's pr


cc @Eikix

File: crates/evm/src/instructions/push_operations.cairo (hunk 2 above)
Comment (enitrat) [reply crates/evm/src/instructions/push_operations.cairo:18]:
```suggestion
        context.stack.push(load_word(i.into(), data))
```

File: crates/evm/src/instructions/push_operations.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/instructions/push_operations.cairo:28]: why not pushing 0 onto the stack directly then

File: crates/evm/src/instructions/push_operations.cairo [hunk 5]
```diff
@@ -1,227 +1,272 @@
...
+        let mut i: usize = i.into();
```
Comment (enitrat) [inline crates/evm/src/instructions/push_operations.cairo:17]:
```suggestion
```

can be deleted if change is made on line below

File: crates/evm/src/instructions/push_operations.cairo (hunk 1 above)
Comment (enitrat) [reply crates/evm/src/instructions/push_operations.cairo:28]: Yeah I think it's better as it's more explicit

File: crates/evm/src/interpreter.cairo [hunk 6]
```diff
@@ -343,135 +343,135 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         }
         if opcode == 95 {
             // PUSH0
-            push_operations::exec_push0(ref context);
+            return context.exec_push0();
```
Comment (enitrat) [inline crates/evm/src/interpreter.cairo:346]: nice

**Review by enitrat — APPROVED** [review]
lgtm, just directly push 0 and we can merge

**Pull request merged**